sql: Add txn instrumentation helper

kyle-a-wong opened the pull request:
Adds a new txnInstrumentationHelper struct for managing
the collection of transaction diagnostic bundles. This
struct has APIs for determining if diagnostics collection
should be started and should continue, to add a statement
bundle to the transaction diagnostics, and to finalize
a transaction diagnostics collection.

Epic: [CRDB-53541](https://cockroachlabs.atlassian.net/browse/CRDB-53541)
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/152886)
<!-- Reviewable:end -->

kyle-a-wong committed:
sql: Add txn instrumentation helper

Adds a new txnInstrumentationHelper struct for managing
the collection of transaction diagnostic bundles. This
struct has APIs for determining if diagnostics collection
should be started and should continue, to add a statement
bundle to the transaction diagnostics, and to finalize
a transaction diagnostics collection.

Part of: CRDB-5342
Epic: CRDB-53541
Release note: None

dhartunian reviewed (commented):
> @dhartunian reviewed 1 of 1 files at r1, 1 of 3 files at r2, 2 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152886)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/sql/stmtdiagnostics/txn_diagnostics.go` line 58 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZG9D2d9QlbDEGyY3Qc:-OZG9D2d9QlbDEGyY3Qd:b-nikw78) ([raw file](https://github.com/cockroachdb/cockroach/blob/07d1afa3baf2ee24ee66b0e953cc9cea6f067f89/pkg/sql/stmtdiagnostics/txn_diagnostics.go#L58)):*
> ```Go
> // as a transaction diagnostic bundle
> type TxnDiagnostic struct {
> 	stmtDiagnostics []StmtDiagnostic
> ```

does this need to include the trace?
___
*[`pkg/sql/stmtdiagnostics/txn_diagnostics_test.go` line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZGCMHU-10TMHmBA37j:-OZGCMHU-10TMHmBA37k:b-au8ol8) ([raw file](https://github.com/cockroachdb/cockroach/blob/07d1afa3baf2ee24ee66b0e953cc9cea6f067f89/pkg/sql/stmtdiagnostics/txn_diagnostics_test.go#L30)):*
> ```Go
> 	defer log.Scope(t).Close(t)
> 
> 	t.Run("request_found", func(t *testing.T) {
> ```

can these tests be table-ized?
___
*[`pkg/sql/stmtdiagnostics/txn_diagnostics_test.go` line 172 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZGD9Ff9fcC-T0FhGxs:-OZGD9Ff9fcC-T0FhGxt:bi9qnft) ([raw file](https://github.com/cockroachdb/cockroach/blob/07d1afa3baf2ee24ee66b0e953cc9cea6f067f89/pkg/sql/stmtdiagnostics/txn_diagnostics_test.go#L172)):*
> ```Go
> 		)
> 		require.Error(t, err)
> 		require.Contains(t, err.Error(), "got non-zero sampling probability")
> ```

you can use `require.ErrorContains` and just have a single check


<!-- Sent from Reviewable.io -->

kyle-a-wong reviewed (commented):
Rebased with master now that the transaction diagnostics PR has been merged. I added trace support to the txn diagnostic collector, which required some additional changes to transaction diagnostics and statement diagnostics, so those changes are also included in this PR. PTAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-:-OZLxXUh04pLbt0AuJnx:basbhvc)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[`pkg/sql/stmtdiagnostics/txn_diagnostics.go` line 58 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZG9D2d9QlbDEGyY3Qc:-OZLxJwSE_nbmRgN-VKs:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/07d1afa3baf2ee24ee66b0e953cc9cea6f067f89/pkg/sql/stmtdiagnostics/txn_diagnostics.go#L58)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

does this need to include the trace?
</blockquote></details>

done
___
*[`pkg/sql/stmtdiagnostics/txn_diagnostics_test.go` line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZGCMHU-10TMHmBA37j:-OZLxM7_8Ur-3-9b_q_7:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/07d1afa3baf2ee24ee66b0e953cc9cea6f067f89/pkg/sql/stmtdiagnostics/txn_diagnostics_test.go#L30)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

can these tests be table-ized?
</blockquote></details>

done
___
*[`pkg/sql/stmtdiagnostics/txn_diagnostics_test.go` line 172 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZGD9Ff9fcC-T0FhGxs:-OZLxQzo21ZZz-WZSzM9:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/07d1afa3baf2ee24ee66b0e953cc9cea6f067f89/pkg/sql/stmtdiagnostics/txn_diagnostics_test.go#L172)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

you can use `require.ErrorContains` and just have a single check
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Nice! I have some drive-by nits.

> @yuzefovich reviewed 1 of 5 files at r3, 7 of 7 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-:-OZctKBABaE3rGy6OS5n:bnlsg7u)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @rytaft)

___
*[`pkg/sql/txn_instrumentation.go` line 49 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZckpQA001vtzhwFXwQ:-OZckpQA001vtzhwFXwR:bd7p9s6) ([raw file](https://github.com/cockroachdb/cockroach/blob/7846761ae57643ad1227f327f9c707d631f717bb/pkg/sql/txn_instrumentation.go#L49)):*
> ```Go
> func (tds *txnDiagnosticsCollector) AddStatementBundle(
> 	stmtFingerprintId uint64, stmt tree.Statement, bundle stmtdiagnostics.StmtDiagnostic,
> ) (bool, error) {
> ```

nit: would be good to add a name to the boolean return parameter.
___
*[`pkg/sql/txn_instrumentation.go` line 82 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZcmhPkBvKmlzWEmMBF:-OZcmhPkBvKmlzWEmMBG:bed68v4) ([raw file](https://github.com/cockroachdb/cockroach/blob/7846761ae57643ad1227f327f9c707d631f717bb/pkg/sql/txn_instrumentation.go#L82)):*
> ```Go
> 	}
> 
> 	id := appstatspb.StmtFingerprintID(tds.request.TxnFingerprintId())
> ```

nit: is this txn fingerprint ID different from `appstatspb.TransactionFingerprintID`?
___
*[`pkg/sql/txn_instrumentation.go` line 142 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZcnV6T7AovtRcYc17k:-OZcnV6T7AovtRcYc17l:b8ib1dv) ([raw file](https://github.com/cockroachdb/cockroach/blob/7846761ae57643ad1227f327f9c707d631f717bb/pkg/sql/txn_instrumentation.go#L142)):*
> ```Go
> 		if err != nil {
> 			log.Ops.VWarningf(ctx, 2, "Failed to add statement bundle: %s", err)
> 			h.Reset()
> ```

nit: this Reset is redundant with the unconditional one below.
___
*[`pkg/sql/stmtdiagnostics/txn_diagnostics.go` line 36 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZcsRlt5W6Nc75Nz4R1:-OZcsRluETV1Ehf9O-Uc:b-9xoxei) ([raw file](https://github.com/cockroachdb/cockroach/blob/7846761ae57643ad1227f327f9c707d631f717bb/pkg/sql/stmtdiagnostics/txn_diagnostics.go#L36)):*
> ```Go
> func NewTxnRequest(
> 	txnFingerprintId uint64,
> 	stmtFingerprintsId []uint64,
> ```

nit: `s/stmtFingerprintsId/stmtFingerprintIds/`.


<!-- Sent from Reviewable.io -->

kyle-a-wong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152886)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @rytaft)

___
*[`pkg/sql/txn_instrumentation.go` line 49 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZckpQA001vtzhwFXwQ:-OZeRvGb5iglwyvZbxdQ:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7846761ae57643ad1227f327f9c707d631f717bb/pkg/sql/txn_instrumentation.go#L49)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: would be good to add a name to the boolean return parameter.
</blockquote></details>

done
___
*[`pkg/sql/txn_instrumentation.go` line 82 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZcmhPkBvKmlzWEmMBF:-OZeRhctApekequnuRXK:b-jbgods) ([raw file](https://github.com/cockroachdb/cockroach/blob/7846761ae57643ad1227f327f9c707d631f717bb/pkg/sql/txn_instrumentation.go#L82)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: is this txn fingerprint ID different from `appstatspb.TransactionFingerprintID`?
</blockquote></details>

Oh yeah. This should be `TransactionFingerprintID`. Also changed this to be the hex string representation of the ID too.
___
*[`pkg/sql/txn_instrumentation.go` line 142 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZcnV6T7AovtRcYc17k:-OZeRsqYCGcFv7n8gsZ2:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7846761ae57643ad1227f327f9c707d631f717bb/pkg/sql/txn_instrumentation.go#L142)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: this Reset is redundant with the unconditional one below.
</blockquote></details>

done
___
*[`pkg/sql/stmtdiagnostics/txn_diagnostics.go` line 36 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZcsRlt5W6Nc75Nz4R1:-OZeRwka1ORScg2-n28x:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7846761ae57643ad1227f327f9c707d631f717bb/pkg/sql/stmtdiagnostics/txn_diagnostics.go#L36)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/stmtFingerprintsId/stmtFingerprintIds/`.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
LGTM. No serious concerns, just some nits and questions.

> @dhartunian reviewed 2 of 7 files at r5, 1 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-:-OZkT3YK0j6xUxs8tWdW:b-y95lwf)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @RaduBerinde, @rytaft, @srosenberg, and @yuzefovich)

___
*[`pkg/sql/stmtdiagnostics/txn_diagnostics_test.go` line 333 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZkKeN5BfrBaZZkrk6F:-OZkKeN5BfrBaZZkrk6G:b4hhj49) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b44f8017e6e63d88ce8aa1b21eaa93b46019042/pkg/sql/stmtdiagnostics/txn_diagnostics_test.go#L333)):*
> ```Go
> 
> 	var bundleChunkCount int
> 	// Verify that statement_diagnostics table now has entries
> ```

nit it's "statement_bundle_chunks" that's being checked.
___
*[`pkg/sql/txn_instrumentation.go` line 63 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZkN9-35QxfZOo5VwWw:-OZkN9-35QxfZOo5VwWx:by3falw) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b44f8017e6e63d88ce8aa1b21eaa93b46019042/pkg/sql/txn_instrumentation.go#L63)):*
> ```Go
> 			return false, errors.Newf("Expected commit or rollback, got %s", stmt)
> 		}
> 		tds.readyToFinalize = true
> ```

a bit confused here: if we rollback does that mean this is ready to finalize? would we finalize this and then start a new bundle?
___
*[`pkg/sql/txn_instrumentation.go` line 77 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZkMkT_6rXzgG8gKzf0:-OZkMkT_6rXzgG8gKzf1:bv329fj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b44f8017e6e63d88ce8aa1b21eaa93b46019042/pkg/sql/txn_instrumentation.go#L77)):*
> ```Go
> 	traceJSONStr, err := tracing.TraceToJSON(trace)
> 	if err != nil {
> 		tds.z.AddFile("trace.json", err.Error())
> ```

nit: should we use `.err` extension?
___
*[`pkg/sql/txn_instrumentation.go` line 96 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZkMl-K5PT6uIwP8NtF:-OZkMl-K5PT6uIwP8NtG:bxud9sk) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b44f8017e6e63d88ce8aa1b21eaa93b46019042/pkg/sql/txn_instrumentation.go#L96)):*
> ```Go
> 	jaegerJSON, err := trace.ToJaegerJSON("txn-trace", comment, "", true /* indent */)
> 	if err != nil {
> 		tds.z.AddFile("trace-jaeger.txt", err.Error())
> ```

nit: nit: should we use `.err` extension?
___
*[`pkg/sql/txn_instrumentation.go` line 148 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZkSAph6I1FZpF4Bp1C:-OZkSAph6I1FZpF4Bp1D:bshpps7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b44f8017e6e63d88ce8aa1b21eaa93b46019042/pkg/sql/txn_instrumentation.go#L148)):*
> ```Go
> }
> 
> func (h *txnInstrumentationHelper) ShouldContinueDiagnostics(
> ```

what's the relationship between `ShouldContinueDiagnostics` and `DiagnosticsInProgress`? should the former set `requestId` to zero when it fails? or is it the job of the caller to walk away at that point?
___
*[`pkg/sql/txn_instrumentation.go` line 184 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZkNIoK8p4qynoNLWwV:-OZkNIoK8p4qynoNLWwW:bui2xy2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b44f8017e6e63d88ce8aa1b21eaa93b46019042/pkg/sql/txn_instrumentation.go#L184)):*
> ```Go
> 			log.Dev.Infof(ctx, "Inserted transaction diagnostics with id %d", id)
> 		}
> 	}
> ```

should we log an error on the else clause here?


<!-- Sent from Reviewable.io -->

kyle-a-wong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152886)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @herkolategan, @RaduBerinde, @rytaft, @srosenberg, and @yuzefovich)

___
*[`pkg/sql/txn_instrumentation.go` line 63 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZkN9-35QxfZOo5VwWw:-OZnxSqoDH_JTUpLvaxI:b-rbavf8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b44f8017e6e63d88ce8aa1b21eaa93b46019042/pkg/sql/txn_instrumentation.go#L63)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

a bit confused here: if we rollback does that mean this is ready to finalize? would we finalize this and then start a new bundle?
</blockquote></details>

As discussed in slack, rollback (`ROLLBACK TRANSACTION`) still completes the transaction so it is considered complete. We will still collect and persist the bundle and consider the diagnostics request complete
___
*[`pkg/sql/txn_instrumentation.go` line 77 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZkMkT_6rXzgG8gKzf0:-OZnvG8qA-EF79ZZjhI3:br7dvry) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b44f8017e6e63d88ce8aa1b21eaa93b46019042/pkg/sql/txn_instrumentation.go#L77)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: should we use `.err` extension?
</blockquote></details>

done. 
I was using the same file names used in statement bundles, but i can switch to .err to make it more clear.
___
*[`pkg/sql/txn_instrumentation.go` line 96 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZkMl-K5PT6uIwP8NtF:-OZnvIjWCVyTIf-Pxuuq:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b44f8017e6e63d88ce8aa1b21eaa93b46019042/pkg/sql/txn_instrumentation.go#L96)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: nit: should we use `.err` extension?
</blockquote></details>

done.
___
*[`pkg/sql/txn_instrumentation.go` line 148 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZkSAph6I1FZpF4Bp1C:-OZnwiXL-4preLC74N42:b-cg0n0b) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b44f8017e6e63d88ce8aa1b21eaa93b46019042/pkg/sql/txn_instrumentation.go#L148)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

what's the relationship between `ShouldContinueDiagnostics` and `DiagnosticsInProgress`? should the former set `requestId` to zero when it fails? or is it the job of the caller to walk away at that point?
</blockquote></details>

For my original implementation, I planned on the latter, but revisiting this, I think it may be better to have this function reset the the in progress diagnostics if it is going to return false. 

I've updated this to call reset.
___
*[`pkg/sql/txn_instrumentation.go` line 184 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZkNIoK8p4qynoNLWwV:-OZnoVy07gl9KVFPXhdK:b3ahg8j) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b44f8017e6e63d88ce8aa1b21eaa93b46019042/pkg/sql/txn_instrumentation.go#L184)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

should we log an error on the else clause here?
</blockquote></details>

My original intention was that when finishing a transaction in `txnState.finishSQLTxn`, it would always call this function and this is the final check of "should we continue with this bundle or discard it". This means that calling this with `readyToFinalize == true` is a valid state and not an error
___
*[`pkg/sql/stmtdiagnostics/txn_diagnostics_test.go` line 333 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-OZkKeN5BfrBaZZkrk6F:-OZnvKRX3XVkCEuUkbzs:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b44f8017e6e63d88ce8aa1b21eaa93b46019042/pkg/sql/stmtdiagnostics/txn_diagnostics_test.go#L333)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit it's "statement_bundle_chunks" that's being checked.
</blockquote></details>

done.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152886#-:-OZp4ALR3m9KaJHrGdYH:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @herkolategan, @RaduBerinde, @rytaft, @srosenberg, and @yuzefovich)



<!-- Sent from Reviewable.io -->
